### PR TITLE
Fix MefTestHelper

### LINKED ISSUE #2467: Infra: improve MEF tests
### Description

The logic in the MefTestHelper is supposed to check that the composition will fail if any required export is missing, but it doesn't.
The new version in `sonarlint-omnisharp` should do a better job.

#### Notes
The SLVS-version used "Take" to select an increasing number of the required elements from the array. However, that's not a good check.

e.g. all inputs [ A, B, C ]
Take[0] -> tests with [ ] -> should fail
Take[1] -> tests with [ A ] -> should fail
Take[1] -> tests with [ A, B ] -> should fail

However, even if all of the iterations fail as expected, the only thing we can conclude is that C is required. It's possible A and B are optional.

The new `sonarlint-omnisharp` version tests the following:
[ B, C ]
[ A, B ]
[ A, C ]
... and expects failure in each case.

### PR DESCRIPTION
**Pull request by duncanp-sonar**
- was not correctly checking for required exports

Fixes #2467

### CODE REVIEW

File: src/IssueViz.Security.UnitTests/OpenInIDE/Http/OwinPipelineProcessorTests.cs [hunk 1]
```diff
@@ -35,16 +36,52 @@ namespace SonarLint.VisualStudio.IssueVisualization.Security.UnitTests.OpenInIDE
     [TestClass]
     public class OwinPipelineProcessorTests
     {
-        const string Origin = "http://origin";
+        private const string Origin = "http://origin";
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/OpenInIDE/Http/OwinPipelineProcessorTests.cs:39]: Miscellaneous: missing scope.

File: src/TestInfrastructure/Helpers/Exceptions.cs [hunk 2]
```diff
@@ -34,67 +34,9 @@ public static class Exceptions
...
-        {
-            if (action == null)
-            {
-                throw new ArgumentNullException("action");
-            }
-
-            try
-            {
-                action();
-                FluentAssertions.Execution.Execute.Assertion.FailWith("Expected exception " + typeof(TException).FullName);
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/Exceptions.cs:83]: This method was buggy: I don't think it's ever worked properly.

File: src/TestInfrastructure/Helpers/Exceptions.cs [hunk 3]
```diff
@@ -34,67 +34,9 @@ public static class Exceptions
...
-                    FluentAssertions.Execution.Execute.Assertion.FailWith("Expected exception " + typeof(TException).FullName + " but got " + ex.GetType().FullName + "\n" + ex.ToString());
-                }
-
-                // Perform any additional checks that were supplied.
-                if (additionalChecks != null)
-                {
-                    additionalChecks(ex);
-                }
-            }
+            action.Should().Throw<TException>();
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/Exceptions.cs:39]: We don't need any of the other methods that were in this class.

We could get rid of this method too, but we use it in 64 places, so it would be in a different PR.

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 4]
```diff
@@ -103,10 +107,13 @@ public static Export CreateExport<T>(object exportInstance, string contractName
             for (int i = 0; i < requiredExports.Count(); i++)
             {
                 // Try importing when not all of the required exports are available -> exception
-                Exceptions.Expect<Exception>(
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:106]: Switching to use the `FluentAssertions` `Should().Throw` fixes one of errors in `CheckTypeCanBeImported`.

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 5]
```diff
@@ -103,10 +107,13 @@ public static Export CreateExport<T>(object exportInstance, string contractName
...
-                    null,
-                    () => TryCompose(importer, catalog, requiredExports.Take(i)),
-                    checkDerived: true);
+                var subsetOfRequiredExports = new List<Export>(requiredExports);
+                var exportToRemove = subsetOfRequiredExports[i];
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:111]: This fixes the other error in `CheckTypeCanBeImported` i.e. the one described in #2467.

File: src/TypeScript.UnitTests/NodeJSLocator/NodeVersionInfoProviderTests.cs [hunk 6]
```diff
@@ -41,8 +40,7 @@ public class NodeVersionInfoProviderTests
         public void MefCtor_CheckIsExported()
         {
             MefTestHelpers.CheckTypeCanBeImported<NodeVersionInfoProvider, INodeVersionInfoProvider>(
-                MefTestHelpers.CreateExport<INodeLocationsProvider>(),
-                MefTestHelpers.CreateExport<ILogger>());
+                MefTestHelpers.CreateExport<INodeLocationsProvider>());
```
Comment (duncanp-sonar) [inline src/TypeScript.UnitTests/NodeJSLocator/NodeVersionInfoProviderTests.cs:43]: This test was passing incorrectly. The `logger` import is not required.

File: src/IssueViz.UnitTests/Selection/IssueSelectionServiceTests.cs [hunk 7]
```diff
@@ -34,8 +34,7 @@ public class IssueSelectionServiceTests
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            MefTestHelpers.CheckTypeCanBeImported<IssueSelectionService, IIssueSelectionService>(
-                MefTestHelpers.CreateExport<IAnalysisIssueSelectionService>());
+            MefTestHelpers.CheckTypeCanBeImported<IssueSelectionService, IIssueSelectionService>();
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Selection/IssueSelectionServiceTests.cs:37]: This test was passing incorrectly - the type has no required imports.

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 8]
```diff
@@ -23,18 +23,22 @@
...
+using System.Diagnostics;
 using System.Linq;
+using FluentAssertions;
 using Moq;
 
 namespace SonarLint.VisualStudio.Integration.UnitTests
 {
     public static class MefTestHelpers
     {
+        [DebuggerStepThrough]
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:35]: Miscellaneous: simplifies debugging through the tests. MS docs are [here](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debuggerstepthroughattribute?view=net-6.0).

File: src/IssueViz.UnitTests/Editor/LocationTagging/IssueLocationStoreAggregatorTests.cs [hunk 9]
```diff
@@ -36,13 +36,24 @@ namespace SonarLint.VisualStudio.IssueVisualization.UnitTests.Editor.LocationTag
     public class IssueLocationStoreAggregatorTests
     {
         [TestMethod]
-        public void MefCtor_CheckExports()
+        public void MefCtor_CheckIsExported()
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/LocationTagging/IssueLocationStoreAggregatorTests.cs:39]: New test - this type effectively has no required types since it is `ImportMany`.

File: src/IssueViz.UnitTests/Editor/LocationTagging/IssueLocationStoreAggregatorTests.cs [hunk 10]
```diff
@@ -36,13 +36,24 @@ namespace SonarLint.VisualStudio.IssueVisualization.UnitTests.Editor.LocationTag
...
+        {
+            // IIssueLocationStores are ImportMany -> optional
+            MefTestHelpers.CheckTypeCanBeImported<IssueLocationStoreAggregator, IIssueLocationStoreAggregator>();
+        }
+
+        [TestMethod]
+        public void MefCtor_CheckIsExported_LocationStoresAreImported()
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/LocationTagging/IssueLocationStoreAggregatorTests.cs:46]: I modified the existing test to check that the expected `ImportMany` types are actually imported.

File: src/IssueViz.Security.UnitTests/OpenInIDE/Http/OwinPipelineProcessorTests.cs [hunk 11]
```diff
@@ -35,16 +36,52 @@ namespace SonarLint.VisualStudio.IssueVisualization.Security.UnitTests.OpenInIDE
...
         public void MefCtor_CheckIsExported()
         {
+            // IOwinRequestHandlers are ImportMany -> optional
             MefTestHelpers.CheckTypeCanBeImported<OwinPipelineProcessor, IOwinPipelineProcessor>(
-                MefTestHelpers.CreateExport<IEnumerable<IOwinPathRequestHandler>>(Array.Empty<IOwinPathRequestHandler>()),
                 MefTestHelpers.CreateExport<ILogger>());
         }
 
+        [TestMethod]
+        public void MefCtor_CheckIsExported_RequestHandlersAreImported()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/OpenInIDE/Http/OwinPipelineProcessorTests.cs:50]: New test to check that the `ImportMany` request handlers are imported as expected.
This test is logically the same as `MefCtor_CheckIsExported_LocationStoresAreImported` (below).

File: src/IssueViz.UnitTests/Editor/IssueSpanCalculatorTests.cs [hunk 12]
```diff
@@ -46,8 +46,7 @@ public void TestInitialize()
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            MefTestHelpers.CheckTypeCanBeImported<IssueSpanCalculator, IIssueSpanCalculator>(
-                MefTestHelpers.CreateExport<IChecksumCalculator>());
+            MefTestHelpers.CheckTypeCanBeImported<IssueSpanCalculator, IIssueSpanCalculator>();
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/IssueSpanCalculatorTests.cs:49]: This test was passing incorrectly before - the import isn't required.

**Review by ugras-ergun-sonarsource — APPROVED** [review]
